### ISSUE TITLE
Deferrable ExternalTaskSensor cannot wait for an entire DAG

### ISSUE THREAD

**Issue by ecodina** [apache/airflow#34207]
### Apache Airflow version

2.7.1

### What happened

When running an ExternalTaskSensor with `external_task_id=None` and in deferrable mode, the trigger doesn't wait for the entire DAG since it needs a `task_id`. However, the typing suggests this should be possible: https://github.com/apache/airflow/blob/main/airflow/triggers/external_task.py#L63

I believe this is the query that doesn't behave as expected from the typing: https://github.com/apache/airflow/blob/main/airflow/triggers/external_task.py#L135

Moreover, after examining the code, it seems the trigger would also fail if the ExternalTaskSensor was provided with an `external_task_ids` (multiple tasks).

### How to reproduce

1. Create `dag1` which contains a simple task (e.g. sleep for 60 seconds)
2. Create `dag2` which contains an ExternalTaskSensor checking for `dag1` (`external_task_id=None`) and runs in deferrable mode
3. Run them and see how `dag2` times out because it can't find any task.

### Operating System

Ubuntu 22.04.3 LTS

### Deployment

Virtualenv installation

### Anything else

Similarly to #34204 I'd love to submit a PR, but have no available time to properly do so or the knowledge to make sure any of my solutions actually work in all cases.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by samc1213** [comment]
I'm interested in taking this issue

**Comment by samc1213** [comment]
I think [this](https://gist.github.com/samc1213/41d58e37d2acba81b4177a773a6ea628) file causes the issue? I see `dag2` in status `failed` - is this a valid way to reproduce the issue @ecodina ? I'm new to airflow so thanks for your help. How do you know it's timing out instead of just failing?

**Comment by ecodina** [comment]
Yes, this file reproduces the issue. You can try lowering the sleep time (e.g. 10 seconds). You'll then see that the ExternalTask keeps logging `Task is still running, sleeping for 2.0 seconds...` after the sleeping task has long finished. It fails after the trigger's 60 seconds hardcoded timeout.

You can also print the result of `TaskStateTrigger.count_tasks` and check it's always 0.

I opened 3 issues regarding the deferrable ExternalTaskSensor and as @yermalov-here commented in #34204 the problem is that the logic of this deferrable trigger is different from the "normal" one. Looking in retrospective, maybe it'd have been better to open 1 issue with the 3 "bugs" since a total refactor may be needed!

**Comment by samc1213** [comment]
Should we consider consolidating the issues? The conversation in #34204 is much more thorough and the fix should address this problem too

**Comment by ecodina** [comment]
Sorry, OOO until October and I just saw your message. It's probably a good idea to consolidate all the issues! I think they all share the same root-cause.

**Comment by Throne3d** [comment]
Hi there! I just hit into this too - would it be possible to add a note that this isn't yet supported in the meantime? It seems like the Triggerer code for this operator was never set up for a full DAG, for multiple task instance IDs, nor for a full task group, which the non-deferrable sensor is able to support. In my case, we switched a bunch of our sensors over to deferrable=True and several ExternalTaskSensors that expected to sense a whole DAG ended up failing for (initially) unclear reasons.

I don't think I'm familiar enough with the Airflow code to fix this all myself, but I could submit an MR for the meantime that this isn't yet functional, so other people are aware of it!

**Comment by potiuk** [comment]
> Hi there! I just hit into this too - would it be possible to add a note that this isn't yet supported in the meantime? 

Sure - it's marked as "good-first-issue" so anyone (including yourself) can add a note in our docs. It's super-easy. Find the doc you want to add it it in, click "Suggest a change on this page" and PR will be opene where you (or whoever does it) can add appropriate comment - and maintainers will comment on it and help you (or anyone else) to merge it. No big familiarity is needed and you are likely one of the best people in the world to know where to add it and how to formulate it so that people like you would make a good use of it.

Can we count on your help there @Throne3d ?

**Comment by Throne3d** [comment]
I've opened up https://github.com/apache/airflow/pull/35854 to add a note here on the sensor page

**Comment by anasatzemoso** [comment]
After running this [file](https://gist.github.com/samc1213/41d58e37d2acba81b4177a773a6ea628) I noticed that the `wait_dag1` task (which is the ExternalTaskSensor) just keeps printing: Sleeping for 2.0 seconds... Even after setting a timeout, the sensor does not stop or fail, it just continues this loop unless the DAG is manually marked as failed. This feels like the timeout is being ignored entirely.

Expectation:
The sensor should exit or fail gracefully once the timeout is reached - not just keep polling forever.
If `external_dag_id` or `external_task_id` is None, the system should handle it with a clear default behavior.

I'm happy to raise a PR to fix this, but before that, I want to make sure I understand the intended behavior correctly. Could someone confirm what should actually happen with the given case?

**Comment by suman-himanshu** [comment]
Tried reproducing the issue using [file](https://gist.github.com/samc1213/41d58e37d2acba81b4177a773a6ea628) from main branch.

As mentioned by @anasatzemoso observed same behaviour, wait_dag1 task keeps printing `Sleeping for 2.0 seconds... Even after setting a timeout, the sensor does not stop or fail, it just continues this loop unless the DAG is manually marked as failed. This feels like the timeout is being ignored entirely`.

<img width="1822" height="1071" alt="Image" src="https://github.com/user-attachments/assets/800953a0-e1bf-409c-bf92-bef54a2c16c4" />

As per my understanding the system should have some default behaviour for the scenario, Please help us understand on what should be the default behaviour in the above scenario.
cc - @akshayvijayjain

**Comment by ecodina** [comment]
I've just tried the deferrable mode in Airflow 3.1.5 and `apache-airflow-providers-standard` 1.10.0 and it seems it works.

"Parent" dag: has a `BashOperator` that sleeps for 10 minutes.

"Sensor" dag: has 2 `ExternalTaskSensor` tasks, 1 waits for the full Dag, another waits for 1 task.

Without setting `execution_timeout` to the `ExternalTaskSensor`

- It waits for the full Dag
- It waits for the `BashOperator` task.

Setting `execution_timeout=timedelta(minutes=1)`:

- Waiting for the full Dag is cancelled after 1 minute
- Waiting for the `BashOperator` task is cancelled after 1 minute

I believe we can now close this issue.

@suman-himanshu are you still experiencing issues? Are you using above versions?

<details>

<summary>Parent Dag</summary>

```python
from airflow.providers.standard.operators.bash import BashOperator
from airflow.sdk import DAG
from datetime import datetime


default_args = {
    'owner': 'test_user',
    'start_date': datetime(2024, 1, 1),
    'weight_rule': 'downstream',
}

with DAG(
    dag_id='sleeper_pipeline',
    default_args=default_args,
    description='Sleeper Pipeline',
    tags=['default'],
) as dag:

    task_a = BashOperator(
        task_id='task_a',
        bash_command='sleep 600',
    )
```

</details>


<details>

<summary>Sensor Dag</summary>

```python
from airflow.providers.standard.sensors.external_task import ExternalTaskSensor
from airflow.sdk import DAG
from airflow.timetables.interval import CronDataIntervalTimetable
from datetime import datetime, timedelta



default_args = {
    'owner': 'test_user',
    'start_date': datetime(2024, 1, 1),
    'weight_rule': 'downstream',
}

with DAG(
    dag_id='sensor_pipeline',
    default_args=default_args,
    description='Sensor Pipeline',
    catchup=False,
    tags=['default'],
) as dag:



    task_a = ExternalTaskSensor(
        task_id='task_a',
        external_dag_id='sleeper_pipeline',
        execution_date_fn=your_function,  # Add your function
        deferrable=True,
        execution_timeout=timedelta(minutes=1), # Remove if you don't want a timeout
    )

    task_b = ExternalTaskSensor(
        task_id='task_b',
        external_dag_id='sleeper_pipeline',
        external_task_ids=['task_a'],
        priority_weight=0,
        execution_date_fn=your_function,  # Add your function
        deferrable=True,
        execution_timeout=timedelta(minutes=1),  # Remove if you don't want a timeout
    )
```

</details>

**Comment by ecodina** [comment]
After 1 month of having ExternalTaskSensor tasks in Dags in Airflow 3, I can confirm it now works properly. I'm closing this issue.
